tests/lib/date_time/src/main.c

```diff
@@ -378,6 +408,525 @@ void test_date_time_validity(void)
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(20);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(40);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(52);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(12);
+
+	k_sem_take(&date_time_callback_sem, K_MSEC(TEST_CB_WAIT_TIME));
+
+	/* Asynchronous update with previous time at 3.5s */
+	k_sleep(K_MSEC(500));
+	date_time_update_async(NULL);
+	k_sem_take(&date_time_callback_sem, K_MSEC(TEST_CB_WAIT_TIME));
+
+	/* Asynchronous update at 5s*/
+	__mock_nrf_modem_at_scanf_ExpectAndReturn("AT+CCLK?", "+CCLK: \"%u/%u/%u,%u:%u:%u%d", 6);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(24);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(1);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(30);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(20);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(40);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(52);
+
+	k_sleep(K_MSEC(1500));
+	date_time_update_async(NULL);
+	k_sem_take(&date_time_callback_sem, K_MSEC(TEST_CB_WAIT_TIME));
+
+	/* Normal cycle at 8s*/
+	__mock_nrf_modem_at_scanf_ExpectAndReturn("AT+CCLK?", "+CCLK: \"%u/%u/%u,%u:%u:%u%d", 7);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(24);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(1);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(30);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(20);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(40);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(52);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(12);
+
+	k_sem_take(&date_time_callback_sem, K_MSEC(TEST_CB_WAIT_TIME));
+
+	/* Time taken into use from %XTIME notification at 9s */
+	k_sleep(K_MSEC(1000));
+	strcpy(at_notif, "%XTIME: ,\"42900180919421\",\"01\"\r\n");
+	at_monitor_dispatch(at_notif);
+
+	k_sem_take(&date_time_callback_sem, K_MSEC(TEST_CB_WAIT_TIME));
+
+	/* TODO: XTIME doesn't delay normal cycle but date_time_update_async does */
+
+	/* Normal cycle at 11s */
+	__mock_nrf_modem_at_scanf_ExpectAndReturn("AT+CCLK?", "+CCLK: \"%u/%u/%u,%u:%u:%u%d", 7);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(24);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(1);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(30);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(20);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(40);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(52);
+	__mock_nrf_modem_at_scanf_ReturnVarg_uint32(12);
+
+	k_sem_take(&date_time_callback_sem, K_MSEC(TEST_CB_WAIT_TIME));
+}
+
+/** Test all retries failing with different NTP failures. */
+void test_date_time_update_all_retry_ntp_fail(void)
+{
+	date_time_cb_count_expected = 3;
+	test_date_time_cb_data[0].type = DATE_TIME_NOT_OBTAINED;
+	test_date_time_cb_data[0].time_expected = 3000;
+	test_date_time_cb_data[1].type = DATE_TIME_NOT_OBTAINED;
+	test_date_time_cb_data[1].time_expected = 4000;
+	test_date_time_cb_data[2].type = DATE_TIME_NOT_OBTAINED;
+	/* TODO: The 2nd retry is already 4.5 due to async call. Is that intentional? */
+	test_date_time_cb_data[2].time_expected = 4500;
```

Comment:
@glarsennordic, I think this is a bit of a flaw (if not a bug) that `date_time_update_async()` takes one retry cycle. We have configurations that there should be `CONFIG_DATE_TIME_RETRY_COUNT=2` retries with `CONFIG_DATE_TIME_RETRY_INTERVAL_SECONDS=1`. If `date_time_update_async()` is called before all retries are done, it will be counted as one retry.
I'm not sure we need to add code to get around this but in that case I will document it in the code.

Reply:
Yes, in fact retries cannot happen at all for `date_time_update_async` if it is called between regularly scheduled updates.

The fix for both issues unfortunately does require keeping two separate retry counters and two separate schedule blocking flags. One set for `date_time_update_async and associated retries`, and the other set for just normal updates

Or we can document this as a limitation. I'm fine with either, though admittedly you are right that this is not exactly the most intuitive behavior

Reply:
> > Yes, in fact retries cannot happen at all for date_time_update_async if it is called between regularly scheduled updates.
>
> This is actually something I didn't even think about. I'll think about it when I get back to this PR.

Well, I was 100% completely wrong about that. As I mention in the other thread, I got myself completely twisted. I apologize for the inaccuracy here too.

I still think the solution to the problem you initially pointed out here is to keep two separate retry counters, but as you said, its not really a big deal at all